### PR TITLE
GPUNetIO Integration

### PR DESCRIPTION
**Pull request by cwharris**

### CODE REVIEW

File: morpheus/_lib/cmake/libmorpheus.cmake [hunk 1]
```diff
@@ -46,6 +46,19 @@ target_link_libraries(cuda_objs
...
+    src/doca/doca_context.cpp
+    src/doca/doca_rx_pipe.cpp
+    src/doca/doca_rx_queue.cpp
+    src/doca/doca_semaphore.cpp
+    src/doca/rte_context.cpp
+    src/stages/doca_source.cpp
+    src/stages/doca_source_kernels.cu
+  )
+endif()
+
```
Comment (mdemoret-nv) [inline morpheus/_lib/cmake/libmorpheus.cmake:61]: The DOCA and Morpheus components seem to be intertwined in this file. What if we built the DOCA pieces as a static library and then linked against it? That way, all of the DOCA specific settings and link options would be confined to a single library and all we would need to do is optionally link against it.

File: morpheus/_lib/include/morpheus/doca/doca_context.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#define DOCA_ALLOW_EXPERIMENTAL_API
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/doca_context.hpp:20]: Does this need to be set before any DOCA files is included? If so, this would be better as a compiler definition. Or we should have a single header file that contains all of the DOCA headers and sets this option first. Otherwise, down the line this might get forgotten about

File: morpheus/_lib/src/doca/doca_context.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+            auto res = doca_dev_open(dev_list[i], retval);
+            if (res == DOCA_SUCCESS)
+            {
+                doca_devinfo_list_destroy(dev_list);
+                return res;
+            }
+        }
+    }
+
+    // DOCA_LOG_WARN("Matching device not found");
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/doca/doca_context.cpp:124]: Commented out warning message

File: morpheus/_lib/src/doca/doca_context.cpp [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+    RTE_TRY(rte_eth_dev_info_get(port_id, &dev_info));
+    RTE_TRY(rte_eth_dev_configure(port_id, rxq_num, rxq_num, &eth_conf));
+
+    mp = rte_pktmbuf_pool_create("TEST", 8192, 0, 0, MAX_PKT_SIZE, rte_eth_dev_socket_id(port_id));
+    if (mp == nullptr)
+    {
+        // DOCA_LOG_ERR("Failed rte_pktmbuf_pool_create with: %s", rte_strerror(-ret));
+        throw std::exception();
+        // return nullptr;
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/doca/doca_context.cpp:168]: Left over debugging code. Also, we should be including an error message with each exception

File: morpheus/_lib/include/morpheus/doca/doca_context.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+#include <string>
+#include <type_traits>
+
+#define GPU_PAGE_SIZE (1UL << 16)
+
+namespace morpheus::doca {
+
+#pragma GCC visibility push(default)
+
+struct DocaContext
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/doca_context.hpp:38]: Use `MORPHEUS_EXPORT` instead of the GCC visibility push/pop
```suggestion
struct MORPHEUS_EXPORT DocaContext
```

Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/doca_context.hpp:38]: Need a docstring on this class describing what it does

File: morpheus/_lib/include/morpheus/doca/doca_mem.hpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+#include <morpheus/doca/error.hpp>
+
+#include <memory>
+
+#define DOCA_ALLOW_EXPERIMENTAL_API
+
+namespace morpheus::doca {
+
+template <typename T>
+struct DocaMem
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/doca_mem.hpp:31]: Need a docstring on this class describing what it does

File: morpheus/_lib/include/morpheus/doca/doca_rx_pipe.hpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#pragma once
+
+#include <morpheus/doca/doca_context.hpp>
+#include <morpheus/doca/doca_rx_queue.hpp>
+
+#include <memory>
+
+namespace morpheus::doca {
+
+struct DocaRxPipe
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/doca_rx_pipe.hpp:27]: Need a docstring on this class describing what it does

File: morpheus/_lib/include/morpheus/doca/doca_rx_queue.hpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+#include <doca_eth_rxq.h>
+#include <doca_gpunetio.h>
+#include <morpheus/doca/doca_context.hpp>
+#include <morpheus/doca/doca_mem.hpp>
+
+#include <memory>
+
+namespace morpheus::doca {
+
+struct DocaRxQueue
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/doca_rx_queue.hpp:29]: Need a docstring on this class describing what it does

File: morpheus/_lib/include/morpheus/doca/doca_semaphore.hpp [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+#pragma once
+
+#include <morpheus/doca/doca_context.hpp>
+
+#include <memory>
+
+namespace morpheus::doca {
+
+struct DocaSemaphore
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/doca_semaphore.hpp:26]: Need a docstring on this class describing what it does

File: morpheus/_lib/include/morpheus/doca/rte_context.hpp [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace morpheus {
+
+namespace doca {
+
+struct RTEContext
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/rte_context.hpp:24]: Need a docstring on this class describing what it does

File: morpheus/_lib/include/morpheus/stages/doca_source.hpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+#include <morpheus/messages/meta.hpp>
+#include <pymrc/node.hpp>
+
+#include <memory>
+
+namespace morpheus {
+
+#pragma GCC visibility push(default)
+
+class DocaSourceStage : public mrc::pymrc::PythonSource<std::shared_ptr<MessageMeta>>
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/doca_source.hpp:33]: Need a docstring on this class describing what it does

File: morpheus/_lib/include/nvtext/normalize.hpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/nvtext/normalize.hpp:1]: I believe that as of cuDF 23.02, the `nvtext` headers are now included in the install. This file, and the entire `nvtext` folder can be removed.

File: morpheus/_lib/include/morpheus/doca/error.hpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+{
+    throw morpheus::doca_error(std::string{"DOCA error encountered at: " + std::string{file} + ":" +
+                                           std::to_string(line) + ": " + std::to_string(error) + " " +
+                                           std::string(doca_get_error_string(error))});
+}
+
+inline void throw_rte_error(int error, const char* file, unsigned int line)
+{
+    throw morpheus::rte_error(std::string{"RTE error encountered at: " + std::string{file} + ":" +
+                                          std::to_string(line) + ": " + std::to_string(error)});
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/doca/error.hpp:50]: We have a macro `MORPHEUS_CONCAT_STR()` which should be used here. Will simplify this significantly.

File: morpheus/_lib/include/morpheus/stages/doca_source_kernels.hpp [hunk 14]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/doca_source_kernels.hpp:1]: Can this file be moved to `morpheus/doca` instead of in the stages directory?

File: morpheus/_lib/include/morpheus/stages/doca_source.hpp [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+                    std::string const& gpu_pci_address,
+                    std::string const& source_ip_filter);
+
+  private:
+    subscriber_fn_t build();
+
+    std::shared_ptr<morpheus::doca::DocaContext> m_context;
+    std::shared_ptr<morpheus::doca::DocaRxQueue> m_rxq;
+    std::shared_ptr<morpheus::doca::DocaRxPipe> m_rxpipe;
+    std::shared_ptr<morpheus::doca::DocaSemaphore> m_semaphore;
```
Comment (mdemoret-nv) [inline morpheus/_lib/include/morpheus/stages/doca_source.hpp:50]: Forward declare these classes so we can remove all of the `doca` include statements from the header.

File: morpheus/_lib/src/stages/doca_source_kernels.cu [hunk 16]
```diff
@@ -0,0 +1,680 @@
+/*
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/doca_source_kernels.cu:1]: Can we move this file out of the stages directory as well?

File: morpheus/_lib/src/stages/doca_source_kernels.cu [hunk 17]
```diff
@@ -0,0 +1,680 @@
...
+    auto packet_idx = threadIdx.x * PACKETS_PER_THREAD + i;
+
+    if (packet_idx >= packet_count) {
+      continue;
+    }
+
+    auto payload_size = payload_sizes[packet_idx];
+    auto packet_idx_out = payload_capture[i];
+
+    // payload_offsets_out[packet_idx_out] = payload_offsets[i];
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/doca_source_kernels.cu:470]: A lot of commented code in this file

File: morpheus/_lib/src/stages/doca_source.cpp [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+            auto packet_count = packet_count_d.value(rmm::cuda_stream_default);
+
+            if (packet_count == 0)
+            {
+                continue;
+            }
+
+            auto payload_size_total = payload_size_total_d.value(rmm::cuda_stream_default);
+
+            // LOG(INFO) << "packet_count(" << packet_count << ") payload_size_total(" << payload_size_total << ")";
```
Comment (mdemoret-nv) [inline morpheus/_lib/src/stages/doca_source.cpp:167]: If you need optional verbose logging, you can use `VLOG(10)` instead of `LOG(INFO)`

File: morpheus/_lib/tests/CMakeLists.txt [hunk 19]
```diff
@@ -44,6 +44,13 @@ target_link_libraries(test_libmorpheus
     pybind11::embed
 )
 
+target_include_directories(test_libmorpheus
+    PUBLIC
+      /opt/mellanox/doca/include/
+      /opt/mellanox/dpdk/include/dpdk
+      /opt/mellanox/dpdk/include/x86_64-linux-gnu/dpdk
+)
+
```
Comment (mdemoret-nv) [inline morpheus/_lib/tests/CMakeLists.txt:53]: Should be wrapped by an if statement

File: morpheus/cli/commands.py [hunk 20]
```diff
@@ -660,3 +660,5 @@ def post_pipeline(ctx: click.Context, *args, **kwargs):
 add_command("train-ae", "morpheus.stages.preprocess.train_ae_stage.TrainAEStage", modes=AE_ONLY)
 add_command("trigger", "morpheus.stages.general.trigger_stage.TriggerStage", modes=ALL)
 add_command("validate", "morpheus.stages.postprocess.validation_stage.ValidationStage", modes=ALL)
+
+add_command("from-doca", "morpheus.stages.doca.doca_source_stage.DocaSourceStage", modes=NLP_ONLY)
```
Comment (mdemoret-nv) [inline morpheus/cli/commands.py:664]: Wont this cause an exception if someone tries to use this without doca being built?

File: morpheus/modules/__init__.py [hunk 21]
```diff
@@ -15,6 +15,7 @@
 Morpheus module definitions, each module is automatically registered when imported
 """
 
+from morpheus._lib import modules
```
Comment (mdemoret-nv) [inline morpheus/modules/__init__.py:18]: Why was this changed

File: morpheus/stages/doca/doca_source_stage.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+    """
+
+    def __init__(
+        self,
+        c: Config,
+        nic_pci_address: str,
+        gpu_pci_address: str,
+    ):
+
+        super().__init__(c)
```
Comment (mdemoret-nv) [inline morpheus/stages/doca/doca_source_stage.py:70]: We probably need some way of detecting if DOCA was built or not and throwing an error here

File: CMakeLists.txt [hunk 23]
```diff
@@ -34,6 +34,7 @@ option(MORPHEUS_USE_CCACHE "Enable caching compilation results with ccache" OFF)
 option(MORPHEUS_USE_CLANG_TIDY "Enable running clang-tidy as part of the build process" OFF)
 option(MORPHEUS_USE_CONDA "Enables finding dependencies via conda instead of vcpkg. Note: This will disable vcpkg. All dependencies must be installed first in the conda environment" OFF)
 option(MORPHEUS_USE_IWYU "Enable running include-what-you-use as part of the build process" OFF)
+option(MORPHEUS_SUPPORT_DOCA "Whether or not to build doca-related elements of morpheus" OFF)
```
Comment (mdemoret-nv) [inline CMakeLists.txt:37]: Keep these sorted

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
A few things we should discuss:

1. Layout of the Dockerfile
2. If there is a better way to organize the files in `thirdparty/artifacts`?
3. Detangling the DOCA and Morpheus components so the DOCA pieces are consolidated in a single location
4. Leftover debugging or commented code

File: morpheus/modules/__init__.py (hunk 21 above)
Comment (cwharris) [reply morpheus/modules/__init__.py:18]: changed automatically by `fix_all.sh`

File: docker/run_container_dev.sh [hunk 24]
```diff
@@ -38,5 +39,13 @@ fi
...
-docker run --rm -ti ${DOCKER_ARGS} ${DOCKER_EXTRA_ARGS} ${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG} "${@:-bash}"
+docker run \
+    -v /dev/hugepages:/dev/hugepages \
+    -v /usr/src/nvidia-515.65.01/build:/usr/src/nvidia-515.65.01/build \
+    -v /lib/modules/5.4.0-128-generic/:/lib/modules/5.4.0-128-generic/ \
+    --privileged \
+    --rm \
+    -ti \
+    ${DOCKER_ARGS} ${DOCKER_EXTRA_ARGS} \
+    ${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG} "${@:-bash}"
```
[the remark this answers is not in the document]
Comment (cwharris) [reply docker/run_container_dev.sh:49]: actually this was originally required to build doca. I'll need to test and make sure we can remove it.

File: docker/Dockerfile [hunk 25]
```diff
@@ -205,6 +328,36 @@ FROM conda_env_dev as development
...
+    registry=https://registry.npmjs.org/\n\
+    " | tee /root/.npmrc >/dev/null' \
+     && ln -s /usr/local/bin/node /usr/local/bin/nodejs \
+     && ln -s /usr/local/lib/node_modules/npm/bin/npm-cli.js /usr/local/bin/npm \
+     && ln -s /usr/local/lib/node_modules/npm/bin/npx-cli.js /usr/local/bin/npx \
+     # Smoke tests
+     && echo "node version: $(node --version)" \
+     && echo " npm version: $(npm --version)" \
+     && echo "yarn version: $(yarn --version)"
+
```
[the remark this answers is not in the document]
Comment (cwharris) [reply docker/Dockerfile:360]: It was to install `node` before we were installing it with apt. This way was recommended by @trxcllnt .

File: docker/Dockerfile [hunk 26]
```diff
@@ -15,23 +15,44 @@
...
+#    packages.
+#  conda_bld_morpheus: Utility stage used to build and cache the Morpheus
+#    conda package for runtime environments. (should not be extended)
+#  runtime: Full runtime environment with Morpheus installed as a conda
+#    package
+#  development: Full development environment able to build against a Morpheus
+#    source directory.
+#  development_pydbg: Development + debug build of cpython and various GDB
+#    debugging macros.
+# ========================================================================= #
```
[the remark this answers is not in the document]
Comment (cwharris) [reply docker/Dockerfile:38]: Thank @drobison00

File: morpheus/_lib/include/morpheus/stages/doca_source_kernels.hpp (hunk 14 above)
Comment (cwharris) [reply morpheus/_lib/include/morpheus/stages/doca_source_kernels.hpp:1]: This file is specific to the doca source stage. I don't see a more appropriate place for it. Perhaps reorganizing the code in the `stages` directory to allow for more complex stage implementations makes sense, but moving this code in to the more reusable `doca` directory just obfuscates what the code is used for in my opinion.

File: docker/Dockerfile (hunk 25 above)
[the remark this answers is not in the document]
Comment (trxcllnt) [reply docker/Dockerfile:358]: The `.npmrc` is optional, but using node from the official images is definitely recommended :+1:

File: morpheus/stages/doca/doca_source_stage.py (hunk 22 above)
Comment (cwharris) [reply morpheus/stages/doca/doca_source_stage.py:70]: Eventually we'll want to do something like that for optional modules by generating a python file (or copying this file to the appropriate location) on build, but we'd want to give more thought to that than we have time for now.

File: morpheus/cli/commands.py (hunk 20 above)
Comment (cwharris) [reply morpheus/cli/commands.py:664]: Yes, and that's probably ok for the time being.

File: morpheus/_lib/cmake/libmorpheus.cmake (hunk 1 above)
Comment (cwharris) [reply morpheus/_lib/cmake/libmorpheus.cmake:61]: Sounds like a good idea to me.

**Review by mdemoret-nv — APPROVED** [review]
